### PR TITLE
Support different NFT contracts with `dao-voting-cw721-staked`

### LINKED ISSUE #716: `dao-voting-cw721-staked` migration
In #692, I made some changes for `dao-voting-cw721-staked` that people may wish to migrate too (support for active threshold as an example).

### PR DESCRIPTION
**Pull request by JakeHartnell**
Now works with `sg721` as well as some custom `cw721` contracts. :partying_face: 

Closes #716

### CODE REVIEW

File: contracts/voting/dao-voting-cw721-staked/src/contract.rs [hunk 1]
```diff
@@ -110,28 +131,25 @@ pub fn instantiate(
             INITITIAL_NFTS.save(deps.storage, &initial_nfts)?;
 
             // Create instantiate submessage for NFT roles contract
```
Comment (noahsaso) [inline contracts/voting/dao-voting-cw721-staked/src/contract.rs]: i think this comment is wrong now

File: contracts/voting/dao-voting-cw721-staked/src/testing/tests.rs [hunk 2]
```diff
@@ -867,6 +916,40 @@ fn test_active_threshold_percentage_lte_0() {
...
+                initial_nfts: vec![to_binary(&Empty {}).unwrap()],
+            },
+            unstaking_duration: None,
+            active_threshold: None,
+        },
+        &[],
+        "cw721_voting",
+        None,
+    )
+    .unwrap_err();
```
Comment (noahsaso) [inline contracts/voting/dao-voting-cw721-staked/src/testing/tests.rs]: can you validate which specific error this returns

**Review by noahsaso — CHANGES_REQUESTED** [review]
looks goooood to me, we should test it on both stargaze and juno testnets before merging

File: contracts/voting/dao-voting-cw721-staked/src/contract.rs [hunk 3]
```diff
@@ -89,10 +89,31 @@ pub fn instantiate(
...
+            // instantiate message. We need to deserialize it and update the minter to be this
+            // contract.
+            //
+            // Custom NFT contracts with Instantiate messages that do fit the cw721 or sg721
+            // interface will not work with this contract. Use NftContract::Existing instead.
+            let cw721_instantiate_msg = from_binary::<cw721_base::InstantiateMsg>(&instantiate_msg);
+            let sg721_instantiate_msg = from_binary::<sg721::InstantiateMsg>(&instantiate_msg);
+
+            // Update the minter to be this contract, reserialize into Binary
+            let instantiate_msg = match (cw721_instantiate_msg, sg721_instantiate_msg) {
```
Comment (bekauz) [inline contracts/voting/dao-voting-cw721-staked/src/contract.rs]: feels like having some sort of structure here could be nice and would help keep the key contract methods short and concise. what do you think?
didn't test but i'm thinking something along the lines of:
```rust
enum NftInstantiateMsg {
    Cw721(cw721_base::InstantiateMsg),
    Sg721(sg721::InstantiateMsg),
    // any future 721 implementations?
}

impl NftInstantiateMsg {
    fn update_minter(&mut self, minter: &str) {
        match self {
            NftInstantiateMsg::Cw721(msg) => msg.minter = minter.to_string(),
            NftInstantiateMsg::Sg721(msg) => msg.minter = minter.to_string(),
        }
    }
}

fn try_deserialize(instantiate_msg: Binary) -> Result<NftInstantiateMsg, ContractError> {
    if let Ok(cw721_msg) = from_binary::<cw721_base::InstantiateMsg>(&instantiate_msg) {
        return Ok(NftInstantiateMsg::Cw721(cw721_msg));
    }

    if let Ok(sg721_msg) = from_binary::<sg721::InstantiateMsg>(&instantiate_msg) {
        return Ok(NftInstantiateMsg::Sg721(sg721_msg));
    }
    
    Err(ContractError::NftInstantiateError {})
}

```

Comment (JakeHartnell) [reply contracts/voting/dao-voting-cw721-staked/src/contract.rs]: Oh, love this.

**Review by noahsaso — APPROVED** [review]
hell yes this looks so good!!

**Pull request merged**